lnwallet/channel.go

```diff
@@ -5410,7 +5184,7 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) (
 
 	remoteHTLCs := lc.channelState.RemoteCommitment.Htlcs
 
-	return fwdPkg, addsToForward, settleFailsToForward, remoteHTLCs, nil
+	return fwdPkg, remoteHTLCs, nil
```

Comment:
I think `addsToForward` and `addUpdates` are different - previously `ReceiveRevocation` would return `addsToForward` and it's then used by other callers, whereas `fwdPkg.Adds` is referencing `addUpdates`. Is this the intended change?

Reply:
I'm glad you're taking a look at this because this is the crux of what I'm not 100% sure about. However, here's my argument for why it's safe.

If you examine [these conditions](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go#L5229-L5245) which is what was required to add it to those payment descriptor slices that were returned, you find that the conditions that allow things to be forwarded are the following:

For adds:
- it is committed to the local commit chain
- it is committed to the remote commit chain
- its remote commit height is *exactly* equal to the remote tail height
- the local commit height is at most the local tail height

For removes:
Same as above but all the add heights are replaced with remove heights.

Now if we consider what the definition of a forward package is, it is the set of net-new htlc adds and "settlefails" that have been fully locked in, which means that we have a valid commitment transaction that commits to them and in this very function call we are receiving the revocation for their commitment transaction that doesn't include these updates. So *as far as I can tell*, there is no reason we should be using anything other than the forwarding package data as the canonical source of truth of what to forward.

Please check me on this, but I *believe* it is correct.

CC @Roasbeef

Reply:
Based on your analysis it seems like `fwdPkg` should be referencing `addsToForward` instead of `addUpdates`?

> So as far as I can tell, there is no reason we should be using anything other than the forwarding package data as the canonical source of truth of what to forward.

Yeah I agree, also love how we are pushing a single source of truth here. When I look at this change, I ask myself three questions,
1. how are `addsToForward` and `addUpdates` created differently?
2. how is `addsToForward` used?
3. how is `fwdPkg.Add` (`addUpdates`) used?

For the first one - `addsToForward` is a subset of `addUpdates`, aka the locked-in updates from all the updates we've received. Will find out the answers for the latter two tomorrow.

Reply:
> Based on your analysis it seems like fwdPkg should be referencing addsToForward instead of addUpdates

I believe that these have a direct mapping between one another. I believe the only difference is that they are different types: `LogUpdate` vs `PaymentDescriptor`

Reply:
I don't think they are the same. `addUpdates` is a list of updates we've received, it's basically a copy of adds found in `remoteUpdateLog`, whereas `addsToForward` is a list of updates that are locked in by this revocation.

Aside from the context, just from this method, `addsToForward` is filtered using `!isForwarded`, `committedAdd`
and `shouldFwdAdd`, but  `addUpdates` just use`!isForwarded`.

I also think we may find a bug here - I think `addsToForward` should be mapped to `fwdPkg.Add` instead of `addUpdates`.

It's also helpful to understand the lifecycle of a `FwdPkg`,

- Create `FwdPkg`: `NewFwdPkg` is called inside `RecieveRevocation`, with `removeChainTail` and all updates (add/settle/fail). This package is saved on disk in `AdvanceCommitChainTail`. `fwdPkg` has state `FwdStateLockedIn`. All the locked-in HTLCs will be set in `fwdPkg.FwdFilter` inside `processRemoteAdds`. These HTLCs will be made into packets and handled by the switch. The `FwdFilter` will be saved on disk.

- Reload `FwdPkg`: The `FwdPkg` is loaded via `loadFwdPkg` and its state is,
  - `FwdStateLockedIn` if `FwdFilter` is nil, which means `processRemoteAdds` was not finished successfully.
  - FwdStateProcessed if there are filters saved, but not full.
  - `FwdStateCompleted` if the saved filters are full, which means all the updates have been forwarded. This package will be GCed.

- Process `FwdPkg.Adds`: When `AckFilter` is not full, we call `processRemoteAdds` with `FwdPkg.Adds`, which assumes all these HTLCs are locked in. We will then use `fwdPkg.AckFilter` to filter out HTLCs that are acked and process the rest.

Given `AckFilter` is updated via `AckAddHtlcs`, I now need to understand the flow there...and yeah we def need this refactor...

Reply:
Ok yeah from what you're saying I haven't seen any showstopper on this just yet. Mostly because the only adds in the remote log that aren't part of the forward package are ones that have already been forwarded. So it ends up being the same set.

Reply:
So `AckAddHtlcs` happens in two methods,
- `AppendRemoteCommitChain`, which happens when we decide to `SignNextCommitment`.
- `cleanupSpuriousResponse`, which is called when an htlc cannot be found when handling the `UpdateFulFillHTLC` and `UpdateFailHTLC`. This htlc will be marked as acked in `ackFilter`.

Seems unrelated to this change.

Since the remote can send us updates and a `revoke_and_ack` that only locks in a subset of the updates, I think we should use `addsToForward` in `fwdPkg.Add` instead of `addUpdates`. I think it means we may have sent unacked updates during the restart, tho it's a rare event.

Reply:
> Since the remote can send us updates and a revoke_and_ack that only locks in a subset of the updates, I think we should use addsToForward in fwdPkg.Add instead of addUpdates. I think it means we may have sent unacked updates during the restart, tho it's a rare event.

I'm not sure what the action item here is. Is the change I made wrong? Or are you asking for me to also include a change in how we compute the forwarding package to begin with?

Reply:
left comments on where I think the change should happen, also wish someone could check me on this to be sure @Roasbeef @Crypt-iQ